File: phase 1/01 - two pointers/lc763.py
def partitionLabels(s):
    # find length and do base case
    slen = len(s)
    if slen<=1:
        return [slen]

    # find the last appearance of each letter
    output = []
    letters = set(s)
    last = {}

    for letter in letters:
        last[letter] = s.rindex(letter)

    #define pointers
    lp = 0
    rp = 0

    #main loop
    for i in range(slen):
        cur = s[i]
        rp = max(rp, last[s[i]])
        if i==rp:
            size = rp-lp+1
            output.append(size)
            rp +=1
            lp = rp

    return output

print(partitionLabels('ababcbacadefegdehijhklij'))

File: phase 1/01 - two pointers/lc763.js
/**
 * @param {string} s
 * @return {number[]}
 */
var partitionLabels = function(s) {

    let slen = s.length;
    if (slen <= 1) {
        return [slen]; // Return an array containing the length of the string
    }

    //find the last appearance of each character
    let letters = new Set(s)
    let last = {}
    let output = []

    for (let letter of letters){
        last[letter] = s.lastIndexOf(letter)
    }
    
    let lp = 0
    let rp = 0

    for (let i = 0; i<slen; i++){
        let curletter = s[i]
        rp = Math.max(rp, last[curletter])
        if (rp == i){
            size = rp-lp+1
            output.push(size)
            rp++
            lp = rp
        }
        
    }

    return output

};

console.log(partitionLabels('ababcbacadefegdehijhklij'))